### PR TITLE
[x264] enable cross-compile settings propagation for arm64-linux

### LINKED ISSUE #22899: [x264] arm64 cross-compile settings are ignored.
**Describe the bug**
[x264] port does **not** propagate cross-compile settings to the 'configure' script. The package is built using host settings leading to output executable format mismatch (if expected output is aarch64 and host is x86_64, for ex).

**Host Environment**
- OS: Linux
- Arch: x86_64
- Compiler: gcc 10.3.0
- Cross-Compiler: linaro arm64 cross-compiler v7.5.0

**Target Environment**
- OS: Linux
- Arch: arm64
- Compiler: gcc linaro v7.5.0

**To Reproduce**

1. Use triplet with any cross-compile toolchain (via VCPKG_CHAINLOAD_TOOLCHAIN_FILE)
2. Add vcpkg-cmake to vcpkg.json
3. Add following lines to portfile
```
vcpkg_cmake_get_vars(cmake_vars_file)
include("${cmake_vars_file}")
```
4. Build
5. Verify that VCPKG_DETECTED_CMAKE_CROSSCOMPILING is set to TRUE in cmake-vars log file.
6. Verify in build log that cross-compile toolchain compiler was **NOT** used.

**Expected behavior**
Build should work, but use the appropriate cross-compile compiler as set per the toolchain.

**Failure logs** 
n.a

**Additional context**
The x264 'configure' options '--host' and '--cross-prefix' are not propagated, hence the build does not use the appropriate tools.

### PR DESCRIPTION
**Pull request by sylvain-prevost**
**Describe the pull request**

- #### What does your PR fix?  
  Fixes #22899 
  (follow-up of #22900)
  Result of X-compilation successfully tested on Jetson nano device.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  !(arm & windows), No

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

**Comment by JackBoosY** [comment]
Depends on https://github.com/microsoft/vcpkg/pull/23525.

**Comment by JackBoosY** [comment]
Close and reopen this PR to re-trigger the ci test.

**Review by JackBoosY — DISMISSED** [review]

**Comment by JackBoosY** [comment]
cc @dg0yt for review this.

File: ports/x264/portfile.cmake [hunk 1]
```diff
@@ -29,7 +29,24 @@ if(VCPKG_TARGET_IS_UWP)
...
+
+    if (VCPKG_TARGET_ARCHITECTURE STREQUAL "arm64")
+
+        vcpkg_cmake_get_vars(cmake_vars_file)
+        include("${cmake_vars_file}")
+
+        if (VCPKG_DETECTED_CMAKE_CROSSCOMPILING STREQUAL "TRUE")
+            # Attempt to determine the fully qualified prefix from VCPKG_DETECTED_CMAKE_C_COMPILER (using most common naming conventions for linux targets)
+            # and if valid, use it to initialize --cross-prefix build-option of x264 package.
+            if (${VCPKG_DETECTED_CMAKE_C_COMPILER} MATCHES ".*-gcc")
```
Comment (dg0yt) [inline ports/x264/portfile.cmake:43]: `VCPKG_DETECTED_CMAKE_C_COMPILER` (or `"${VCPKG_DETECTED_CMAKE_C_COMPILER}"`)

File: ports/x264/portfile.cmake [hunk 2]
```diff
@@ -29,7 +29,24 @@ if(VCPKG_TARGET_IS_UWP)
 endif()
 
 if(VCPKG_TARGET_IS_LINUX)
-    list(APPEND OPTIONS --enable-pic)
+
+    list(APPEND OPTIONS --enable-pic)
```
Comment (dg0yt) [inline ports/x264/portfile.cmake:33]: Remove trailing whitespace.

File: ports/x264/portfile.cmake [hunk 3]
```diff
@@ -29,7 +29,24 @@ if(VCPKG_TARGET_IS_UWP)
...
-    list(APPEND OPTIONS --enable-pic)
+
+    list(APPEND OPTIONS --enable-pic)    
+
+    if (VCPKG_TARGET_ARCHITECTURE STREQUAL "arm64")
+
+        vcpkg_cmake_get_vars(cmake_vars_file)
+        include("${cmake_vars_file}")
+
+        if (VCPKG_DETECTED_CMAKE_CROSSCOMPILING STREQUAL "TRUE")
```
Comment (dg0yt) [inline ports/x264/portfile.cmake:40]: `if (VCPKG_DETECTED_CMAKE_CROSSCOMPILING)`

File: ports/x264/portfile.cmake [hunk 4]
```diff
@@ -29,7 +29,24 @@ if(VCPKG_TARGET_IS_UWP)
...
+            # Attempt to determine the fully qualified prefix from VCPKG_DETECTED_CMAKE_C_COMPILER (using most common naming conventions for linux targets)
+            # and if valid, use it to initialize --cross-prefix build-option of x264 package.
+            if (${VCPKG_DETECTED_CMAKE_C_COMPILER} MATCHES ".*-gcc")
+                string(REGEX REPLACE "gcc$" "" full_cross_prefix ${VCPKG_DETECTED_CMAKE_C_COMPILER})
```
Comment (dg0yt) [inline ports/x264/portfile.cmake:44]: `"${VCPKG_DETECTED_CMAKE_C_COMPILER}"`

File: ports/x264/portfile.cmake [hunk 5]
```diff
@@ -29,7 +29,24 @@ if(VCPKG_TARGET_IS_UWP)
...
+                list(APPEND OPTIONS --cross-prefix=${full_cross_prefix})
```
Comment (dg0yt) [inline ports/x264/portfile.cmake:45]: `"--cross-prefix=${full_cross_prefix}"`

File: ports/x264/vcpkg.json [hunk 6]
```diff
@@ -1,14 +1,19 @@
 {
   "name": "x264",
-  "version-string": "164-5db6aa6cab1b146",
-  "port-version": 4,
+  "version": "164-5db6aa6cab1b146",
```
Comment (dg0yt) [inline ports/x264/vcpkg.json:3]: `version` :  For dot-separated numeric versions 
https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/manifest-files.md#version-fields

File: ports/x264/vcpkg.json [hunk 7]
```diff
@@ -1,14 +1,19 @@
...
+  "port-version": 5,
   "description": "x264 is a free software library and application for encoding video streams into the H.264/MPEG-4 AVC compression format",
   "homepage": "https://github.com/mirror/x264",
+  "license": "GPL-2.0",
```
Comment (dg0yt) [inline ports/x264/vcpkg.json:7]: Some headers say "either version 2 of the License, or (at your option) any later version". But there are also files with other licenses.
My first reference (after the sources is debian copyright):
https://tracker.debian.org/media/packages/x/x264/copyright-20.163.3060git5db6aa6-2

**Review by dg0yt — CHANGES_REQUESTED** [review]
Disclaimer: I don't use this port.

File: ports/x264/portfile.cmake (hunk 1 above)
Comment (dg0yt) [reply ports/x264/portfile.cmake:43]: The regex should be `"-gcc\$"`.
The whole if block could be rewritten as (untested):
~~~cmake
if(VCPKG_DETECTED_CMAKE_C_COMPILER MATCHES "^(.*-)gcc\$")
    list(APPEND OPTIONS "--cross-prefix=${CMAKE_MATCH_1}")
endif()

File: ports/x264/vcpkg.json (hunk 7 above)
Comment (sylvain-prevost) [reply ports/x264/vcpkg.json:7]: Agreed. was going with the 'Copying' file content & statement on https://www.videolan.org/developers/x264.html, but truth is in the code itself which reference the above statement in nearly every file...   so going to change it to GPL2-or-later :)

File: ports/x264/vcpkg.json (hunk 6 above)
Comment (sylvain-prevost) [reply ports/x264/vcpkg.json:3]: That's how it was !? 
Changed it to please CI and corresponding warning that it should be modified.. On account that CI was not behaving properly, I'll revert and see if warning comes back again.

File: ports/x264/portfile.cmake (hunk 1 above)
Comment (sylvain-prevost) [reply ports/x264/portfile.cmake:43]: tested it. works as expected.

**Comment by sylvain-prevost** [comment]
Currently fails on PR formatting due to 'version-string'. 
If 'version' is used instead of 'version-string', it passes the checks but then it does not match the dot-version definition. 
Any clue/guidance ?

**Comment by dg0yt** [comment]
> Currently fails on PR formatting due to 'version-string'.
> If 'version' is used instead of 'version-string', it passes the checks but then it does not match the dot-version definition.
> Any clue/guidance ?

Created https://github.com/microsoft/vcpkg/issues/23585.

**Comment by JackBoosY** [comment]
Can you please rerun the command `,/vcpkg format-manifest ports/x264/vcpkg.json` then check the changes?

**Comment by sylvain-prevost** [comment]
Tried to re-trig the CI.. but looks like same behavior..
Do I need to update vcpkg tool ?

**Comment by dg0yt** [comment]
Try
~~~
git commit --allow-empty -m CI
git push
~~~

**Comment by sylvain-prevost** [comment]
Looks like there is a new error on x64-windows build (which does not seem directly related).


-- Configuring x64-windows
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:128 (message):
    Command failed: ninja -v
    Working Directory: D:/buildtrees/qtinterfaceframework/x64-windows-rel/vcpkg-parallel-configure
    Error code: 1
    See logs for more information:
      D:\buildtrees\qtinterfaceframework\config-x64-windows-out.log

Call Stack (most recent call first):
  D:/installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:415 (vcpkg_execute_required_process)
  D:/installed/x64-windows/share/qtbase/qt_install_submodule.cmake:102 (vcpkg_cmake_configure)
  ports/qtinterfaceframework/portfile.cmake:84 (qt_cmake_configure)
  scripts/ports.cmake:145 (include)

Error: Building package qtinterfaceframework:x64-windows failed with: BUILD_FAILED

**Comment by LilyWangLL** [comment]
cc @Neumann-A Could you help take a look the baseline error? `qtinterfaceframework:x64-windows` build failed with the following error:
```
-- Running ifcodegen for D:/buildtrees/qtinterfaceframework/src/564c6bf227-ee28b9e83e.clean/src/ifvehiclefunctions/ifvehiclefunctions.qface with template frontend
CMake Error at src/interfaceframework/Qt6InterfaceFrameworkMacros.cmake:377 (message):
  Error while running the ifcodegen:

  D:/downloads/tools/python/python-3.10.2-x64/python.exe
  D:/buildtrees/qtinterfaceframework/src/564c6bf227-ee28b9e83e.clean/src/tools/ifcodegen/generate.py
  -T
  D:/buildtrees/qtinterfaceframework/src/564c6bf227-ee28b9e83e.clean/src/tools/ifcodegen/templates
  --template=frontend --force --module=QtIfVehicleFunctions
  D:/buildtrees/qtinterfaceframework/src/564c6bf227-ee28b9e83e.clean/src/ifvehiclefunctions/ifvehiclefunctions.qface
  D:/buildtrees/qtinterfaceframework/x64-windows-rel/src/ifvehiclefunctions

  Traceback (most recent call last):

    File "D:\buildtrees\qtinterfaceframework\src\564c6bf227-ee28b9e83e.clean\src\tools\ifcodegen\generate.py", line 40, in <module>
      from qface.generator import FileSystem, Generator
    File "D:\downloads\tools\python\python-3.10.2-x64\lib\site-packages\qface\generator.py", line 24, in <module>
      from .filters import get_filters
    File "D:\downloads\tools\python\python-3.10.2-x64\lib\site-packages\qface\filters.py", line 3, in <module>
      from .helper import qtcpp
    File "D:\downloads\tools\python\python-3.10.2-x64\lib\site-packages\qface\helper\qtcpp.py", line 5, in <module>
      from jinja2 import environmentfilter

  ImportError: cannot import name 'environmentfilter' from 'jinja2'
  (D:\downloads\tools\python\python-3.10.2-x64\lib\site-packages\jinja2\__init__.py)


Call Stack (most recent call first):
  src/interfaceframework/Qt6InterfaceFrameworkMacros.cmake:391 (qt6_ifcodegen_generate)
  src/ifvehiclefunctions/CMakeLists.txt:8 (qt_ifcodegen_generate)
```

**Comment by sylvain-prevost** [comment]
issue #23766 is not anymore present, however the version-string issue is still there :(

When I created the PR, I used the command

`vcpkg x-add-version x264`

and the output was:

`Use the version scheme "version" instead of "version-string" in port "x264".`
`Use `--skip-version-format-check` to disable this check.`

Thereafter I used the suggested flag to be able to generate the version & baseline  files.
Since then there is this PR Suggestion CI build error that is there.

If I replace 'version-string' by 'version' then it passes the check, but that would be invalid since the value of the version is: ''164-5db6aa6cab1b146" and this is not a dot-version.

**Comment by JackBoosY** [comment]
Oh I totally lost track with this PR, sorry.

**Comment by strega-nil-ms** [comment]
I don't know enough about this to say - I'm marking this as requires:discussion so we can talk with the team on thursday. Thanks for the PR!

**Comment by strega-nil-ms** [comment]
This seems wrong; can you please post the `CC=` line from `buildtrees/x264/arm64-linux-dbg/config.mak` and the `VCPKG_DETECTED_CMAKE_C_COMPILER` line from `buildtrees/x264/cmake-vars-arm64-linux-dbg.cmake.log`, before the change that you've made? That should let us isolate where the failure is.

**Comment by sylvain-prevost** [comment]
Hi @LilyWangLL ,

1. To be able to generate the config.mak file, the first thing is to enable option --disable-asm for linux & arm64.

2. Once done, here are the requested lines:

**from buildtrees/x264/[target]-rel/config.mak:** 
`CC=gcc`

**from buildtrees/x264/cmake-vars-[target]-rel.cmake.log:** 
`set(VCPKG_DETECTED_CMAKE_C_COMPILER "/root/builds/jetson-nano-cross-compile-assets/toolchains/toolchain-gcc-cross-linaro-7.5.0/bin/aarch64-linux-gnu-gcc")`

Just in case, to reduce ping-pong, I'm adding in attachment the corresponding files.

[config.mak.log](https://github.com/microsoft/vcpkg/files/8567967/config.mak.log)
 
[cmake-vars-jetson-rel.cmake.log](https://github.com/microsoft/vcpkg/files/8567965/cmake-vars-jetson-rel.cmake.log)

**Comment by sylvain-prevost** [comment]
And the triplet file I'm using for my target (jetson nano), cross-compiled on WSL2 Ubuntu 20.04:

set (VCPKG_TARGET_ARCHITECTURE          arm64)
set (VCPKG_CRT_LINKAGE                  dynamic)
set (VCPKG_LIBRARY_LINKAGE              static)
set (VCPKG_CMAKE_SYSTEM_NAME            Linux)
set (VCPKG_CMAKE_SYSTEM_VERSION         Tegra)

set (VCPKG_MAKE_BUILD_TRIPLET           --host=aarch64-linux-gnu)

if (DEFINED ENV{GSTREAMER_FOR_DEEPSTREAM})
    # gstreamer is required to be built as shared library (for nVidia deepstream), along with few other libraries.
    if (${PORT} MATCHES "gstreamer|pango|cairo|glib|pcre|libffi")
        set (VCPKG_LIBRARY_LINKAGE      dynamic)
    endif()
endif()

set (VCPKG_CHAINLOAD_TOOLCHAIN_FILE      ${CMAKE_CURRENT_LIST_DIR}/../toolchains/aarch64_linux_jetson.cmake)

**Comment by sylvain-prevost** [comment]
@strega-nil-ms , requested info is available.

File: ports/x264/portfile.cmake [hunk 8]
```diff
@@ -29,7 +29,23 @@ if(VCPKG_TARGET_IS_UWP)
 endif()
 
 if(VCPKG_TARGET_IS_LINUX)
+
     list(APPEND OPTIONS --enable-pic)
+
+    if (VCPKG_TARGET_ARCHITECTURE STREQUAL "arm64")
```
Comment (BillyONeal) [inline ports/x264/portfile.cmake:35]: This condition is probably wrong: Why is cross compiling for arm64 different than any other cross compiling?

**Review by BillyONeal — CHANGES_REQUESTED** [review]
As in https://github.com/microsoft/vcpkg/pull/22331/ , ideally we would be setting all the right tools to the right bits (via `CC`, `AR`, etc.) rather than trying to coax x264's build scripts into detecting the right bits themselves.

Based on https://github.com/mirror/x264/blob/master/configure#L557 it looks like you need:

* `CC`
* `STRINGS`
* `AR`
* `RANLIB`
* `PKGCONFIG` (??? it seems like this runs on the host so we don't understand why it is cross-prefix'd)
* `STRIP` (but we don't enable stripping so this one may not matter)

**Comment by BillyONeal** [comment]
>Based on https://github.com/mirror/x264/blob/master/configure#L557 it looks like you need:

To clarify, you need to get these values from `vcpkg_cmake_get_vars` and set those environment variables as consumed by that `configure` script. Start with `CC`, `AR`, and `RANLIB` since with those it will probably just work.

**Comment by JackBoosY** [comment]
Convert this PR to draft since there is no progress.

**Comment by JackBoosY** [comment]
Closing this PR since it seems that no progress is being made. Please reopen if work is still being done.

**Pull request closed without merging**